mycroft/configuration/locations.py

```diff
@@ -13,11 +13,16 @@
 # limitations under the License.
 import os
 from os.path import join, dirname, expanduser, exists
+import xdg
 
 DEFAULT_CONFIG = join(dirname(__file__), 'mycroft.conf')
 SYSTEM_CONFIG = os.environ.get('MYCROFT_SYSTEM_CONFIG',
                                '/etc/mycroft/mycroft.conf')
-USER_CONFIG = join(expanduser('~'), '.mycroft/mycroft.conf')
+# Make sure we support the old location still
+# Deprecated and will be removed eventually
+OLD_USER_CONFIG = join(expanduser('~'), '.mycroft/mycroft.conf')
+USER_CONFIG = join(xdg.BaseDirectory.save_config_path('mycroft'),
```

Comment:
a string definition should not create a folder, especially by simply importing this, other places would be ok-ish

eg, if every service launches at the exact same time there is a millisecond window where 2 services try to create the directory and one of them throws an exception and fails to load, this sounds hard to cause but i can replicate consistently in my setup

Reply:
In that case it should be replaced with `xdg.BaseDirectory.load_first_config('mycroft')`.

Reply:
what if config does not exist? what does that return?

the main usage outside of core for this is to edit the user config, `LocalConf(USER_CONFIG).store()`, this is what creates the file if it does not exist, so shouldnt it be always the same place? if this operation creates different files it gets messy. 

`USER_CONFIG = join(xdg.BaseDirectory.xdg_config_home, BASE_FOLDER, CONFIG_FILE_NAME)`

this does not impact loading of other xdg dirs, but ensures the string points to a valid deterministic path, its kinda like the other places where you have a read location and a write location

Reply:
https://github.com/takluyver/pyxdg/pull/15 also fixes this, but i still think its not clean to create a directory on a string definition, doesnt feel like the proper place. eg, if i want to check if that directory exists it gets created on import, so i can't do that check

Reply:
Why would you ever want to check if that directory exists though? Is there _any_ use-case for _not_ having the config folder present? Who cares that it's created on string definition. You just ask for the location of the folder, and you get it. With pyxdg you can safely assume it exists, which is fine as long as there is no use-case for not having the folder.

Reply:
in OVOS we do some checks like these to determine in what platform we are running for example, so yes, there are use edge cases like that. For instance since this PR has been open for over a year i have used this to determine if xdg is supported at all, this is part of what allows ovos_utils to know its running in the alpine patched version of mycroft-core but not on HolmesV ;)

Reply:
> You are welcome to disagree with me

I definitely disagree here lol :wink: 

> in OVOS we do some checks like these to determine in what platform we are running for example

You check _what platform_ you are running on by checking if this folder exists? Or just if XDG is supported or not? If the latter, that would definitely be the only use-case ever for that check. I really see no reason to ever check if that directory exists in a pyxdg codebase as it's safe to assume the directory exists if you use the pyxdg save method to get the path (which you should).

Reply:
getting a path != creating a path

let's just agree to disagree, the race condition issue stands and your suggestion wont work, we can also block this PR until a new release of pyxdg with PR linked above

> You check what platform you are running on by checking if this folder exists? Or just if XDG is supported or not?

its a datapoint in a larger heuristic for platform detection, OVOS approach is more capability based than hardcoded list of platform strings

Reply:
duh, disregard what i said, i forgot about `__ensure_folder_exists`, you can use that to workaround the race condition mentioned above, and update it with same fix from pyxdg PR, it has the same issue